### ISSUE TITLE
GPU memory leak 

### ISSUE THREAD

**Issue by arunpatala** [ncullen93/torchsample#27]
Hi, 

I was using this repo for training my networks in pytorch and it has been very helpful and made the code concise. But I was facing out of memory issues while training for long number of iterations or epochs while training on GPUs. When the training starts the usage is around half. As the iterations happen, memory usage is becoming full. Please help.

Thanks

**Comment by benwu232** [comment]
I met similar problem while using predict_loader. Still debugging.

**Comment by arunpatala** [comment]
I have defaulted the tensor to Cuda. Changing that seems to have fixed the issue

**Comment by ncullen93** [comment]
Hey sorry I've been out of town. Are you using regularizers or constraints?

**Comment by arunpatala** [comment]
I think train and val are using separate memory for input variables. May be they should be merged in a single loop

**Comment by recastrodiaz** [comment]
@ncullen93 I experienced the same issue as @arunpatala when running predict() and predict_loader() methods on large GPU datasets. A fix I found is to move all output predictions to cpu() before running torch.cats(). I have implemented this fix in one of my branches, happy to make a PR if you like the idea: https://github.com/recastrodiaz/torchsample/blob/19e80210868632597e14ba400b4ff72ac17e5697/torchsample/modules/module_trainer.py#L424

Whilst training, I have also experienced RAM Out Of Memory errors when using fit() and shuffle=True. I bypassed this issue by building a [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) and a [DataLoader](http://pytorch.org/docs/master/_modules/torch/utils/data/dataloader.html#DataLoader) with shuffle=True. 

One option would be to make the fit() method internally call the fit_loader() function. This would avoid the need to manually create batches, and make the fit() method more flexible by allowing us to take advantage of the extra DataLoader settings such as shuffle, pin_memory, num_workers, batch_sampler, etc.
